toffy/rosetta_test.py

```diff
@@ -54,12 +57,115 @@ def test_compensate_matrix_simple():
     # last channel in second fov is changed by baseline * 2 * 10 due to fov and coefficient
     assert np.all(out[1, :, :, -1] == inputs[1, :, :, -1] - total_comp * 10 * 2)
 
+    # don't generate output for first channel
+    out_indices = out_indices[1:]
+    out = rosetta.compensate_matrix_simple(inputs, coeffs, out_indices)
+
+    # non-affected channels are identical
+    assert np.all(out[:, :, :, :-1] == inputs[:, :, :, 1:-1])
+
+    # last channel is changed by baseline amount * 2 due to multiplier in coefficient matrix
+    assert np.all(out[0, :, :, -1] == inputs[0, :, :, -1] - total_comp * 2)
+
+    # last channel in second fov is changed by baseline * 2 * 10 due to fov and coefficient
+    assert np.all(out[1, :, :, -1] == inputs[1, :, :, -1] - total_comp * 10 * 2)
+
+
+def test_validate_inputs():
+    with tempfile.TemporaryDirectory() as top_level_dir:
+        data_dir = os.path.join(top_level_dir, 'data_dir')
+        os.makedirs(data_dir)
+
+        # make fake data for testing
+        fovs, chans = test_utils.gen_fov_chan_names(num_fovs=3, num_chans=3)
+        filelocs, data_xr = test_utils.create_paired_xarray_fovs(
+            data_dir, fovs, chans, img_shape=(10, 10), fills=True)
 
+        # generate default, correct values for all parameters
+        masses = [71, 76, 101]
+        comp_mat_vals = np.random.rand(len(masses), len(masses)) / 100
+        comp_mat = pd.DataFrame(comp_mat_vals, columns=masses, index=masses)
+
+        acquired_masses = copy.copy(masses)
+        acquired_targets = copy.copy(chans)
+        input_masses = copy.copy(masses)
+        output_masses = copy.copy(masses)
+        all_masses = copy.copy(masses)
+        save_format = 'raw'
+        batch_size = 1
+        gaus_rad = 1
+
+        input_dict = {'raw_data_dir': data_dir, 'comp_mat': comp_mat,
+                      'acquired_masses': acquired_masses, 'acquired_targets': acquired_targets,
+                      'input_masses': input_masses,
+                      'output_masses': output_masses, 'all_masses': all_masses, 'fovs': fovs,
+                      'save_format': save_format, 'batch_size': batch_size, 'gaus_rad': gaus_rad}
+
+        # check that masses are sorted
+        input_dict_disorder = copy.copy(input_dict)
+        input_dict_disorder['acquired_masses'] = masses[1:] + masses[:1]
+        with pytest.raises(ValueError, match='Masses must be sorted'):
+            rosetta.validate_inputs(**input_dict_disorder)
+
+        # check that all masses are present
+        input_dict_missing = copy.copy(input_dict)
+        input_dict_missing['acquired_masses'] = masses[1:]
+        with pytest.raises(ValueError, match='acquired masses and list compensation masses'):
+            rosetta.validate_inputs(**input_dict_missing)
```

Comment:
I could do a dict that contains an entry for each test, with the key, new value, and error string stored in it, but I think it might be a bit more difficult to read